{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Python Testing Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using PyMySQL version 1.4.6\n",
      "Using PyMongo version 4.10.1\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import pymysql\n",
    "except:\n",
    "    print(\"Missing pymysql import, installing now\")\n",
    "    %pip install pymysql\n",
    "try:\n",
    "    import pymongo\n",
    "except:\n",
    "    print(\"Missing pymongo import, installing now\")\n",
    "    %pip install pymongo\n",
    "\n",
    "import json\n",
    "print(\"Using PyMySQL version\", pymysql.__version__)\n",
    "print(\"Using PyMongo version\", pymongo.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "(2003, \"Can't connect to MySQL server on 'chatdb-sql.c7ewm2wuk3r8.us-east-2.rds.amazonaws.com' ([WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond)\")",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTimeoutError\u001b[0m                              Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymysql\\connections.py:649\u001b[0m, in \u001b[0;36mConnection.connect\u001b[1;34m(self, sock)\u001b[0m\n\u001b[0;32m    648\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 649\u001b[0m     sock \u001b[38;5;241m=\u001b[39m socket\u001b[38;5;241m.\u001b[39mcreate_connection(\n\u001b[0;32m    650\u001b[0m         (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mport), \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnect_timeout, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[0;32m    651\u001b[0m     )\n\u001b[0;32m    652\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\socket.py:853\u001b[0m, in \u001b[0;36mcreate_connection\u001b[1;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[0;32m    852\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m all_errors:\n\u001b[1;32m--> 853\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exceptions[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m    854\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m ExceptionGroup(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcreate_connection failed\u001b[39m\u001b[38;5;124m\"\u001b[39m, exceptions)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\socket.py:838\u001b[0m, in \u001b[0;36mcreate_connection\u001b[1;34m(address, timeout, source_address, all_errors)\u001b[0m\n\u001b[0;32m    837\u001b[0m     sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[1;32m--> 838\u001b[0m sock\u001b[38;5;241m.\u001b[39mconnect(sa)\n\u001b[0;32m    839\u001b[0m \u001b[38;5;66;03m# Break explicitly a reference cycle\u001b[39;00m\n",
      "\u001b[1;31mTimeoutError\u001b[0m: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[8], line 7\u001b[0m\n\u001b[0;32m      4\u001b[0m database_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchatdb-sql\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m      5\u001b[0m port \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m3306\u001b[39m  \u001b[38;5;66;03m# Default MySQL port\u001b[39;00m\n\u001b[1;32m----> 7\u001b[0m connection \u001b[38;5;241m=\u001b[39m pymysql\u001b[38;5;241m.\u001b[39mconnect(\n\u001b[0;32m      8\u001b[0m     host\u001b[38;5;241m=\u001b[39mendpoint,\n\u001b[0;32m      9\u001b[0m     user\u001b[38;5;241m=\u001b[39musername,\n\u001b[0;32m     10\u001b[0m     password\u001b[38;5;241m=\u001b[39mpassword,\n\u001b[0;32m     11\u001b[0m     database\u001b[38;5;241m=\u001b[39mdatabase_name,\n\u001b[0;32m     12\u001b[0m     connect_timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m100\u001b[39m\n\u001b[0;32m     13\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymysql\\connections.py:361\u001b[0m, in \u001b[0;36mConnection.__init__\u001b[1;34m(self, user, password, host, database, unix_socket, port, charset, collation, sql_mode, read_default_file, conv, use_unicode, client_flag, cursorclass, init_command, connect_timeout, read_default_group, autocommit, local_infile, max_allowed_packet, defer_connect, auth_plugin_map, read_timeout, write_timeout, bind_address, binary_prefix, program_name, server_public_key, ssl, ssl_ca, ssl_cert, ssl_disabled, ssl_key, ssl_key_password, ssl_verify_cert, ssl_verify_identity, compress, named_pipe, passwd, db)\u001b[0m\n\u001b[0;32m    359\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    360\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 361\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnect()\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymysql\\connections.py:716\u001b[0m, in \u001b[0;36mConnection.connect\u001b[1;34m(self, sock)\u001b[0m\n\u001b[0;32m    714\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m DEBUG:\n\u001b[0;32m    715\u001b[0m         \u001b[38;5;28mprint\u001b[39m(exc\u001b[38;5;241m.\u001b[39mtraceback)\n\u001b[1;32m--> 716\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[0;32m    718\u001b[0m \u001b[38;5;66;03m# If e is neither DatabaseError or IOError, It's a bug.\u001b[39;00m\n\u001b[0;32m    719\u001b[0m \u001b[38;5;66;03m# But raising AssertionError hides original error.\u001b[39;00m\n\u001b[0;32m    720\u001b[0m \u001b[38;5;66;03m# So just reraise it.\u001b[39;00m\n\u001b[0;32m    721\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "\u001b[1;31mOperationalError\u001b[0m: (2003, \"Can't connect to MySQL server on 'chatdb-sql.c7ewm2wuk3r8.us-east-2.rds.amazonaws.com' ([WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond)\")"
     ]
    }
   ],
   "source": [
    "endpoint = \"chatdb-sql.c7ewm2wuk3r8.us-east-2.rds.amazonaws.com\"\n",
    "username = \"admin\"\n",
    "password = \"DSCICh4tDBP455\"\n",
    "database_name = \"chatdb-sql\"\n",
    "port = 3306  # Default MySQL port\n",
    "\n",
    "connection = pymysql.connect(\n",
    "    host=endpoint,\n",
    "    user=username,\n",
    "    password=password,\n",
    "    database=database_name,\n",
    "    connect_timeout=100\n",
    ")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Collection0']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "mongo_username = 'mdmolnar'\n",
    "mongo_password = 'AtM0nG0d1452'\n",
    "# mongo_username = 'jacobgra'\n",
    "# mongo_password = 'AtM0nG0d145'\n",
    "\n",
    "\n",
    "connection_string = f'mongodb+srv://{mongo_username}:{mongo_password}@cluster0.tgu2d.mongodb.net/'\n",
    "\n",
    "client = pymongo.MongoClient(connection_string)\n",
    "\n",
    "# Access a database\n",
    "db = client[\"ChatDB\"]\n",
    "collection0 = db[\"Collection0\"]\n",
    "collection1 = db[\"Collection1\"]\n",
    "\n",
    "# Test connection by listing collections\n",
    "print(db.list_collection_names())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "BulkWriteError",
     "evalue": "batch op errors occurred, full error: {'writeErrors': [{'index': 0, 'code': 11000, 'errmsg': 'E11000 duplicate key error collection: ChatDB.Collection0 index: _id_ dup key: { _id: \"605c72d4bcf86cd799439022\" }', 'keyPattern': {'_id': 1}, 'keyValue': {'_id': '605c72d4bcf86cd799439022'}, 'op': {'_id': '605c72d4bcf86cd799439022', 'name': 'Wireless Headphones', 'description': 'Bluetooth wireless headphones with noise cancellation.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 99.99, 'stock': 100, 'brand': 'BrandX', 'createdAt': {'$date': '2023-10-01T10:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439011'}, 'rating': 4}, {'userId': {'$oid': '605c72d4bcf86cd799439012'}, 'rating': 5}]}}], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 0, 'nModified': 0, 'nRemoved': 0, 'upserted': []}",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mBulkWriteError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 8\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;66;03m# If the JSON data is an array of documents, use insert_many\u001b[39;00m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(file_data, \u001b[38;5;28mlist\u001b[39m):\n\u001b[1;32m----> 8\u001b[0m     collection0\u001b[38;5;241m.\u001b[39minsert_many(file_data)\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m     10\u001b[0m     collection0\u001b[38;5;241m.\u001b[39minsert_one(file_data)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\_csot.py:119\u001b[0m, in \u001b[0;36mapply.<locals>.csot_wrapper\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    117\u001b[0m         \u001b[38;5;28;01mwith\u001b[39;00m _TimeoutContext(timeout):\n\u001b[0;32m    118\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m func(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m--> 119\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m func(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\synchronous\\collection.py:975\u001b[0m, in \u001b[0;36mCollection.insert_many\u001b[1;34m(self, documents, ordered, bypass_document_validation, session, comment)\u001b[0m\n\u001b[0;32m    973\u001b[0m blk \u001b[38;5;241m=\u001b[39m _Bulk(\u001b[38;5;28mself\u001b[39m, ordered, bypass_document_validation, comment\u001b[38;5;241m=\u001b[39mcomment)\n\u001b[0;32m    974\u001b[0m blk\u001b[38;5;241m.\u001b[39mops \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(gen())\n\u001b[1;32m--> 975\u001b[0m blk\u001b[38;5;241m.\u001b[39mexecute(write_concern, session, _Op\u001b[38;5;241m.\u001b[39mINSERT)\n\u001b[0;32m    976\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m InsertManyResult(inserted_ids, write_concern\u001b[38;5;241m.\u001b[39macknowledged)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\synchronous\\bulk.py:736\u001b[0m, in \u001b[0;36m_Bulk.execute\u001b[1;34m(self, write_concern, session, operation)\u001b[0m\n\u001b[0;32m    734\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m    735\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 736\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mexecute_command(generator, write_concern, session, operation)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\synchronous\\bulk.py:603\u001b[0m, in \u001b[0;36m_Bulk.execute_command\u001b[1;34m(self, generator, write_concern, session, operation)\u001b[0m\n\u001b[0;32m    593\u001b[0m _ \u001b[38;5;241m=\u001b[39m client\u001b[38;5;241m.\u001b[39m_retryable_write(\n\u001b[0;32m    594\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_retryable,\n\u001b[0;32m    595\u001b[0m     retryable_bulk,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    599\u001b[0m     operation_id\u001b[38;5;241m=\u001b[39mop_id,\n\u001b[0;32m    600\u001b[0m )\n\u001b[0;32m    602\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m full_result[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwriteErrors\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01mor\u001b[39;00m full_result[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwriteConcernErrors\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m--> 603\u001b[0m     _raise_bulk_write_error(full_result)\n\u001b[0;32m    604\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m full_result\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\bulk_shared.py:131\u001b[0m, in \u001b[0;36m_raise_bulk_write_error\u001b[1;34m(full_result)\u001b[0m\n\u001b[0;32m    125\u001b[0m         errmsg \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m    126\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThis MongoDB deployment does not support \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    127\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mretryable writes. Please add retryWrites=false \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    128\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto your connection string.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    129\u001b[0m         )\n\u001b[0;32m    130\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m OperationFailure(errmsg, code, full_result)\n\u001b[1;32m--> 131\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m BulkWriteError(full_result)\n",
      "\u001b[1;31mBulkWriteError\u001b[0m: batch op errors occurred, full error: {'writeErrors': [{'index': 0, 'code': 11000, 'errmsg': 'E11000 duplicate key error collection: ChatDB.Collection0 index: _id_ dup key: { _id: \"605c72d4bcf86cd799439022\" }', 'keyPattern': {'_id': 1}, 'keyValue': {'_id': '605c72d4bcf86cd799439022'}, 'op': {'_id': '605c72d4bcf86cd799439022', 'name': 'Wireless Headphones', 'description': 'Bluetooth wireless headphones with noise cancellation.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 99.99, 'stock': 100, 'brand': 'BrandX', 'createdAt': {'$date': '2023-10-01T10:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439011'}, 'rating': 4}, {'userId': {'$oid': '605c72d4bcf86cd799439012'}, 'rating': 5}]}}], 'writeConcernErrors': [], 'nInserted': 0, 'nUpserted': 0, 'nMatched': 0, 'nModified': 0, 'nRemoved': 0, 'upserted': []}"
     ]
    }
   ],
   "source": [
    "path_to_json = '../data/orders_json/products.json'\n",
    "# upload '../gyg' to collection1\n",
    " \n",
    "with open(path_to_json) as file:\n",
    "    file_data = json.load(file)\n",
    "\n",
    "# If the JSON data is an array of documents, use insert_many\n",
    "if isinstance(file_data, list):\n",
    "    collection0.insert_many(file_data)\n",
    "else:\n",
    "    collection0.insert_one(file_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'_id': '605c72d4bcf86cd799439026', 'name': 'Tablet', 'description': '10-inch tablet with high-resolution display and long battery life.', 'category': {'$oid': '605c72d4bcf86cd799439034'}, 'price': 499.99, 'stock': 80, 'brand': 'BrandB', 'createdAt': {'$date': '2023-09-30T15:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439016'}, 'rating': 5}]}\n",
      "{'_id': '605c72d4bcf86cd799439024', 'name': 'Smartwatch', 'description': 'Fitness and activity tracker with heart rate monitoring.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 199.99, 'stock': 150, 'brand': 'BrandZ', 'createdAt': {'$date': '2023-10-04T11:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439013'}, 'rating': 4}]}\n",
      "{'_id': '605c72d4bcf86cd799439022', 'name': 'Wireless Headphones', 'description': 'Bluetooth wireless headphones with noise cancellation.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 99.99, 'stock': 100, 'brand': 'BrandX', 'createdAt': {'$date': '2023-10-01T10:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439011'}, 'rating': 4}, {'userId': {'$oid': '605c72d4bcf86cd799439012'}, 'rating': 5}]}\n",
      "{'_id': '605c72d4bcf86cd799439025', 'name': 'Laptop', 'description': 'High-performance laptop for gaming and work.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 1299.99, 'stock': 30, 'brand': 'BrandA', 'createdAt': {'$date': '2023-09-25T09:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439014'}, 'rating': 5}, {'userId': {'$oid': '605c72d4bcf86cd799439015'}, 'rating': 4}]}\n",
      "{'_id': '605c72d4bcf86cd799439023', 'name': 'Smartphone', 'description': 'Latest model with advanced features.', 'category': {'$oid': '605c72d4bcf86cd799439034'}, 'price': 799.99, 'stock': 50, 'brand': 'BrandY', 'createdAt': {'$date': '2023-10-03T14:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439012'}, 'rating': 5}]}\n",
      "{'_id': '605c72d4bcf86cd799439028', 'name': '4K TV', 'description': '65-inch 4K Ultra HD TV with HDR support.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 999.99, 'stock': 20, 'brand': 'BrandD', 'createdAt': {'$date': '2023-09-15T18:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439018'}, 'rating': 5}]}\n",
      "{'_id': '605c72d4bcf86cd799439029', 'name': 'Digital Camera', 'description': '24.2MP digital camera with 4K video recording capabilities.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 699.99, 'stock': 60, 'brand': 'BrandE', 'createdAt': {'$date': '2023-09-18T13:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439019'}, 'rating': 4}]}\n",
      "{'_id': '605c72d4bcf86cd799439027', 'name': 'Bluetooth Speaker', 'description': 'Portable Bluetooth speaker with 360-degree sound.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 49.99, 'stock': 200, 'brand': 'BrandC', 'createdAt': {'$date': '2023-09-20T12:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439011'}, 'rating': 3}, {'userId': {'$oid': '605c72d4bcf86cd799439017'}, 'rating': 4}]}\n",
      "{'_id': '605c72d4bcf86cd799439030', 'name': 'Wireless Charger', 'description': 'Fast wireless charger for smartphones and accessories.', 'category': {'$oid': '605c72d4bcf86cd799439034'}, 'price': 29.99, 'stock': 500, 'brand': 'BrandF', 'createdAt': {'$date': '2023-09-12T10:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439020'}, 'rating': 4}]}\n",
      "{'_id': '605c72d4bcf86cd799439031', 'name': 'Gaming Console', 'description': 'Next-gen gaming console with 1TB storage.', 'category': {'$oid': '605c72d4bcf86cd799439033'}, 'price': 499.99, 'stock': 40, 'brand': 'BrandG', 'createdAt': {'$date': '2023-08-15T11:00:00Z'}, 'ratings': [{'userId': {'$oid': '605c72d4bcf86cd799439011'}, 'rating': 5}, {'userId': {'$oid': '605c72d4bcf86cd799439013'}, 'rating': 4}]}\n"
     ]
    }
   ],
   "source": [
    "for i in collection0.find():\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "filter must be an instance of dict, bson.son.SON, or any other type that inherits from collections.Mapping",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m collection0\u001b[38;5;241m.\u001b[39mfind({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m$id\u001b[39m\u001b[38;5;124m'\u001b[39m})\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\synchronous\\collection.py:1920\u001b[0m, in \u001b[0;36mCollection.find\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1734\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfind\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs: Any, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Cursor[_DocumentType]:\n\u001b[0;32m   1735\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Query the database.\u001b[39;00m\n\u001b[0;32m   1736\u001b[0m \n\u001b[0;32m   1737\u001b[0m \u001b[38;5;124;03m    The `filter` argument is a query document that all results\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1918\u001b[0m \u001b[38;5;124;03m    .. seealso:: The MongoDB documentation on `find <https://dochub.mongodb.org/core/find>`_.\u001b[39;00m\n\u001b[0;32m   1919\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1920\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m Cursor(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\synchronous\\cursor.py:147\u001b[0m, in \u001b[0;36mCursor.__init__\u001b[1;34m(self, collection, filter, projection, skip, limit, no_cursor_timeout, cursor_type, sort, allow_partial_results, oplog_replay, batch_size, collation, hint, max_scan, max_time_ms, max, min, return_key, show_record_id, snapshot, comment, session, allow_disk_use, let)\u001b[0m\n\u001b[0;32m    144\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_explicit_session \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m    146\u001b[0m spec: Mapping[\u001b[38;5;28mstr\u001b[39m, Any] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mfilter\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m {}\n\u001b[1;32m--> 147\u001b[0m validate_is_mapping(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfilter\u001b[39m\u001b[38;5;124m\"\u001b[39m, spec)\n\u001b[0;32m    148\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(skip, \u001b[38;5;28mint\u001b[39m):\n\u001b[0;32m    149\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mskip must be an instance of int\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\mmoln\\anaconda3\\Lib\\site-packages\\pymongo\\common.py:541\u001b[0m, in \u001b[0;36mvalidate_is_mapping\u001b[1;34m(option, value)\u001b[0m\n\u001b[0;32m    539\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Validate the type of method arguments that expect a document.\"\"\"\u001b[39;00m\n\u001b[0;32m    540\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(value, abc\u001b[38;5;241m.\u001b[39mMapping):\n\u001b[1;32m--> 541\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[0;32m    542\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00moption\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m must be an instance of dict, bson.son.SON, or \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    543\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124many other type that inherits from \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    544\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcollections.Mapping\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    545\u001b[0m     )\n",
      "\u001b[1;31mTypeError\u001b[0m: filter must be an instance of dict, bson.son.SON, or any other type that inherits from collections.Mapping"
     ]
    }
   ],
   "source": [
    "collection0.find({'$id'})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
